README.md
# Helix Version Picker

> A preflight service that determines if a repo has opted in to running an upcoming version of helix-pages.

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-version-picker.svg)](https://codecov.io/gh/adobe/helix-version-picker)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-version-picker.svg)](https://circleci.com/gh/adobe/helix-version-picker)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-version-picker.svg)](https://github.com/adobe/helix-version-picker/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-version-picker.svg)](https://github.com/adobe/helix-version-picker/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-version-picker.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-version-picker)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

## Usage

```bash
curl https://adobeioruntime.net/api/v1/web/helix/helix-services/version-picker@v1
```

For more, see the [API documentation](docs/API.md).

## Development

### Deploying Helix Version Picker

Deploying Helix Version Picker requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

All commits to `main` that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/version-picker@ci<num>` and tagged with the CI build number.


CHANGELOG.md
## [1.4.32](https://github.com/adobe/helix-version-picker/compare/v1.4.31...v1.4.32) (2022-02-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v3.0.4 ([4f8d213](https://github.com/adobe/helix-version-picker/commit/4f8d2135c59a9d5b045564257021fa2e4bb28ef3))

## [1.4.31](https://github.com/adobe/helix-version-picker/compare/v1.4.30...v1.4.31) (2022-02-09)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-wrap to v1.0.2 ([#247](https://github.com/adobe/helix-version-picker/issues/247)) ([4379dd9](https://github.com/adobe/helix-version-picker/commit/4379dd9ec8cb85cdc019e1d1b20c4a10463fb3a4))

## [1.4.30](https://github.com/adobe/helix-version-picker/compare/v1.4.29...v1.4.30) (2022-02-04)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.5 ([#241](https://github.com/adobe/helix-version-picker/issues/241)) ([b9def20](https://github.com/adobe/helix-version-picker/commit/b9def20dfa8514666d4cddee301a91f3710e048c))

## [1.4.29](https://github.com/adobe/helix-version-picker/compare/v1.4.28...v1.4.29) (2022-02-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-wrap to v1.0.1 ([#239](https://github.com/adobe/helix-version-picker/issues/239)) ([22ce2ba](https://github.com/adobe/helix-version-picker/commit/22ce2ba839ac32a5b58871029e5b608b5bd58b39))

## [1.4.28](https://github.com/adobe/helix-version-picker/compare/v1.4.27...v1.4.28) (2022-01-10)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v2 ([#225](https://github.com/adobe/helix-version-picker/issues/225)) ([e2706ea](https://github.com/adobe/helix-version-picker/commit/e2706ea005f4b3892df8ae494276a260a737ddee))

## [1.4.27](https://github.com/adobe/helix-version-picker/compare/v1.4.26...v1.4.27) (2021-11-26)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.9.1 ([#211](https://github.com/adobe/helix-version-picker/issues/211)) ([bd70446](https://github.com/adobe/helix-version-picker/commit/bd704460fe02bb39b5f473a140c62e1ff7d95f40))

## [1.4.26](https://github.com/adobe/helix-version-picker/compare/v1.4.25...v1.4.26) (2021-11-22)


### Bug Fixes

* **deps:** pin dependencies ([#209](https://github.com/adobe/helix-version-picker/issues/209)) ([b37ed11](https://github.com/adobe/helix-version-picker/commit/b37ed11d56c2352a8fb6dc4e22d1bf23c89e4e86))

## [1.4.25](https://github.com/adobe/helix-version-picker/compare/v1.4.24...v1.4.25) (2021-10-20)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal-logger to v2 ([#197](https://github.com/adobe/helix-version-picker/issues/197)) ([2dd5c2c](https://github.com/adobe/helix-version-picker/commit/2dd5c2c9923a63397d30176e4cf46c8665627c52))

## [1.4.24](https://github.com/adobe/helix-version-picker/compare/v1.4.23...v1.4.24) (2021-10-19)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.9.0 ([#196](https://github.com/adobe/helix-version-picker/issues/196)) ([4ba0e27](https://github.com/adobe/helix-version-picker/commit/4ba0e27ca8d2ea14aa79d8238d24cf58212fd4e8))

## [1.4.23](https://github.com/adobe/helix-version-picker/compare/v1.4.22...v1.4.23) (2021-10-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v3 ([c5ac760](https://github.com/adobe/helix-version-picker/commit/c5ac7601f4497c3f2b1990095ea29220e5af6347))

## [1.4.22](https://github.com/adobe/helix-version-picker/compare/v1.4.21...v1.4.22) (2021-09-27)


### Bug Fixes

* **deps:** use secrets-manager ([5bb0d8c](https://github.com/adobe/helix-version-picker/commit/5bb0d8c5b38dc229c35b5e56e17ac4e5998d1f3a))

## [1.4.21](https://github.com/adobe/helix-version-picker/compare/v1.4.20...v1.4.21) (2021-09-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.3 ([#182](https://github.com/adobe/helix-version-picker/issues/182)) ([ba9f608](https://github.com/adobe/helix-version-picker/commit/ba9f6085c26cbe1a27a6ed8237cd457cc883b7e6))

## [1.4.20](https://github.com/adobe/helix-version-picker/compare/v1.4.19...v1.4.20) (2021-09-17)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.12 ([#179](https://github.com/adobe/helix-version-picker/issues/179)) ([a217644](https://github.com/adobe/helix-version-picker/commit/a21764471f97dcfce547e12a84964deae1b2a3b6))

## [1.4.19](https://github.com/adobe/helix-version-picker/compare/v1.4.18...v1.4.19) (2021-09-11)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.10 ([#176](https://github.com/adobe/helix-version-picker/issues/176)) ([8bcf511](https://github.com/adobe/helix-version-picker/commit/8bcf511ff6bcbc700e37a18b5dbed6446c22b450))

## [1.4.18](https://github.com/adobe/helix-version-picker/compare/v1.4.17...v1.4.18) (2021-09-10)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.9 ([#174](https://github.com/adobe/helix-version-picker/issues/174)) ([ac57248](https://github.com/adobe/helix-version-picker/commit/ac57248744ebb57345f4ebabdf8128eb4b6a5045))

## [1.4.17](https://github.com/adobe/helix-version-picker/compare/v1.4.16...v1.4.17) (2021-09-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal-logger to v1.0.8 ([b585119](https://github.com/adobe/helix-version-picker/commit/b585119ceac8cf3082c112bba53efa1d9b9813ff))

## [1.4.16](https://github.com/adobe/helix-version-picker/compare/v1.4.15...v1.4.16) (2021-09-02)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.8 ([b4494b7](https://github.com/adobe/helix-version-picker/commit/b4494b77c04948265d10ef73b9ba1e6b6e5b609d))

## [1.4.15](https://github.com/adobe/helix-version-picker/compare/v1.4.14...v1.4.15) (2021-08-19)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.7 ([016e571](https://github.com/adobe/helix-version-picker/commit/016e57165f1aa9c40955524b14833cb7d1e1f9ea))

## [1.4.14](https://github.com/adobe/helix-version-picker/compare/v1.4.13...v1.4.14) (2021-08-18)


### Bug Fixes


test/utils.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const querystring = require('querystring');

function retrofit(fn) {
  const resolver = {
    createURL({ package, name, version }) {
      return new URL(`https://adobeioruntime.net/api/v1/web/helix/${package}/${name}@${version}`);
    },
  };
  return async (params = {}, env = {}) => {
    const context = {
      resolver,
      env,
      // eslint-disable-next-line no-underscore-dangle
      log: params.__ow_logger,
    };
    const resp = await fn({
      url: `https://version-picker.com/version?${querystring.encode(params)}`,
      headers: new Map(Object.entries(params.__ow_headers || {})),
    }, context);
    return {
      statusCode: resp.status,
      body: await resp.text(),
      headers: resp.headers.plain(),
    };
  };
}

module.exports = { retrofit };


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

'use strict';

process.env.HELIX_FETCH_FORCE_HTTP1 = true;

const assert = require('assert');
const nock = require('nock');
const { main } = require('../src/index.js');
const { retrofit } = require('./utils.js');

const index = retrofit(main);

describe('Index Tests', () => {
  afterEach(() => {
    nock.cleanAll();
  });

  it('returns 400 if owner is missing', async () => {
    const result = await index({});
    assert.deepEqual(result, {
      body: 'owner, repo, ref required.',
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'content-type': 'text/plain; charset=utf-8',
      },
      statusCode: 400,
    });
  });

  it('returns 400 if repo is missing', async () => {
    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
      },
    });
    assert.deepEqual(result, {
      body: 'owner, repo, ref required.',
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'content-type': 'text/plain; charset=utf-8',
      },
      statusCode: 400,
    });
  });

  it('returns 400 if ref is missing', async () => {
    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
        'x-repo': 'test-repo',
      },
    });
    assert.deepEqual(result, {
      body: 'owner, repo, ref required.',
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'content-type': 'text/plain; charset=utf-8',
      },
      statusCode: 400,
    });
  });

  it('index function returns the version from github', async () => {
    nock('https://raw.githubusercontent.com')
      .get('/test-owner/test-repo/main/helix-version.txt')
      .reply(200, 'foo-bar');

    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
        'x-repo': 'test-repo',
        'x-ref': 'main',
      },
    });
    assert.deepEqual(result, {
      body: 'foo-bar',
      statusCode: 200,
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'x-pages-version': 'foo-bar',
        'surrogate-control': 'max-age: 30',
        'surrogate-key': 'preflight-main--test-repo--test-owner',
        vary: 'X-Owner,X-Repo,X-Ref,X-Repo-Root-Path',
        'content-type': 'text/plain; charset=utf-8',
      },
    });
  });

  it('index function returns the version from private github', async () => {
    nock('https://raw.githubusercontent.com')
      .get('/test-owner/test-repo/main/helix-version.txt')
      .matchHeader('authorization', 'token ghp_foobar')
      .reply(200, 'foo-bar');

    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
        'x-repo': 'test-repo',
        'x-ref': 'main',
        'x-github-token': 'ghp_foobar',
      },
    });
    assert.deepEqual(result, {
      body: 'foo-bar',
      statusCode: 200,
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'x-pages-version': 'foo-bar',
        'surrogate-control': 'max-age: 30',
        'surrogate-key': 'preflight-main--test-repo--test-owner',
        vary: 'X-Owner,X-Repo,X-Ref,X-Repo-Root-Path',
        'content-type': 'text/plain; charset=utf-8',
      },
    });
  });

  it('index function returns the version from github with alternate root', async () => {
    nock('https://www.example.com')
      .get('/test-owner/test-repo/main/helix-version.txt')
      .reply(200, 'foo-bar');

    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
        'x-repo': 'test-repo',
        'x-ref': 'main',
        'x-repo-root-path': 'https://www.example.com',
      },
    });
    assert.deepEqual(result, {
      body: 'foo-bar',
      statusCode: 200,
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'x-pages-version': 'foo-bar',
        'surrogate-control': 'max-age: 30',
        'surrogate-key': 'preflight-main--test-repo--test-owner',
        vary: 'X-Owner,X-Repo,X-Ref,X-Repo-Root-Path',
        'content-type': 'text/plain; charset=utf-8',
      },
    });
  });

  it('index function returns 404 the version does not exist on github', async () => {
    nock('https://raw.githubusercontent.com')
      .get('/test-owner/test-repo/main/helix-version.txt')
      .reply(404);

    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
        'x-repo': 'test-repo',
        'x-ref': 'main',
      },
    });
    assert.deepEqual(result, {
      body: 'no version',
      statusCode: 200,
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'surrogate-control': 'max-age: 30',
        'surrogate-key': 'preflight-main--test-repo--test-owner',
        vary: 'X-Owner,X-Repo,X-Ref,X-Repo-Root-Path',
        'content-type': 'text/plain; charset=utf-8',
      },
    });
  });

  it('index function returns 504 if github errors', async () => {
    nock('https://raw.githubusercontent.com')
      .get('/test-owner/test-repo/main/helix-version.txt')
      .reply(500);

    const result = await index({
      __ow_headers: {
        'x-owner': 'test-owner',
        'x-repo': 'test-repo',
        'x-ref': 'main',
      },
    });
    assert.deepEqual(result, {
      body: 'unable to fetch version',
      statusCode: 504,
      headers: {
        'cache-control': 'no-store, private, must-revalidate',
        'content-type': 'text/plain; charset=utf-8',
      },
    });
  });
});


test/post-deploy.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */

const chai = require('chai');
const chaiHttp = require('chai-http');
const { createTargets } = require('./post-deploy-utils.js');

chai.use(chaiHttp);
const { expect } = chai;

createTargets().forEach((target) => {
  describe(`Post-Deploy Tests (${target.title()})`, () => {
    it('Returns version', async () => {
      await chai
        .request(target.host())
        .get(target.urlPath())
        .set('X-Owner', 'tripodsan')
        .set('X-Repo', 'helix-pages-test')
        .set('X-Ref', 'picker-unit-test')
        .then((response) => {
          expect(response).to.have.status(200);
          expect(response).to.have.header('x-pages-version', 'breaking-january');
        })
        .catch((e) => {
          throw e;
        });
    }).timeout(50000);
  });
});


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const URL = require('url');
const wrap = require('@adobe/helix-shared-wrap');
const { logger } = require('@adobe/helix-universal-logger');
const { wrap: status } = require('@adobe/helix-status');
const { Response } = require('@adobe/helix-fetch');
const fetchAPI = require('@adobe/helix-fetch');

function createFetchContext() {
  /* istanbul ignore next */
  if (process.env.HELIX_FETCH_FORCE_HTTP1) {
    return fetchAPI.context({ alpnProtocols: [fetchAPI.ALPN_HTTP1_1] });
  }
  /* istanbul ignore next */
  return fetchAPI.context({});
}
const fetchContext = createFetchContext();
const { fetch } = fetchContext;

function computeGithubURI(root, owner, repo, ref, path) {
  const rootURI = URL.parse(root);
  const rootPath = rootURI.path;
  // remove double slashes
  const fullPath = `${rootPath}/${owner}/${repo}/${ref}/${path}`.replace(
    /\/+/g,
    '/',
  );

  rootURI.pathname = fullPath;
  return URL.format(rootURI);
}

/**
 * Generates an error response
 * @param {string} message - error message
 * @param {number} statusCode - error code.
 * @returns response
 */
function error(message, statusCode) {
  return new Response(message, {
    status: statusCode,
    headers: {
      'Cache-Control': 'no-store, private, must-revalidate',
    },
    body: message,
  });
}

async function getVersion(url, githubToken) {
  const fetchopts = {
    cache: 'no-store',
    signal: fetchContext.timeoutSignal(5000),
    'Cache-Control': 'no-cache',
  };
  if (githubToken) {
    fetchopts.headers = {
      authorization: `token ${githubToken}`,
    };
  }
  const resp = await fetch(url, fetchopts);
  const text = await resp.text();
  if (resp.ok) {
    // todo: validate if proper version
    return text.trim();
  }
  if (resp.status !== 404) {
    throw Error(`github error: ${resp.status} ${text}`);
  }
  return '';
}

/**
 * This is the main function
 * @returns {object} a greeting
 */
async function main(request, context) {
  const owner = request.headers.get('x-owner');
  const repo = request.headers.get('x-repo');
  const ref = request.headers.get('x-ref');
  const root = request.headers.get('x-repo-root-path') || 'https://raw.githubusercontent.com/';
  const githubToken = request.headers.get('x-github-token');
  const { log } = context;

  if (!owner || !repo || !ref) {
    log.warn('owner, repo, ref missing');
    return error('owner, repo, ref required.', 400);
  }

  let version = '';

  try {
    const url = computeGithubURI(root, owner, repo, ref, '/helix-version.txt');
    version = await getVersion(url, githubToken);
  } catch (e) {
    log.error('error while fetching version', e);
    return error('unable to fetch version', 504);
  }
  log.info(`version for ${repo}/${owner}#${ref} = "${version}"`);
  const surrogateKey = `preflight-${ref}--${repo}--${owner}`;

  if (version) {
    return new Response(version, {
      status: 200,
      headers: {
        'x-pages-version': version,
        'Cache-Control': 'no-store, private, must-revalidate', // todo: proper caching ??
        'Surrogate-Control': 'max-age: 30',
        'Surrogate-Key': surrogateKey,
        Vary: 'X-Owner,X-Repo,X-Ref,X-Repo-Root-Path',
      },
    });
  }
  return new Response('no version', {
    status: 200,
    headers: {
      'Cache-Control': 'no-store, private, must-revalidate', // todo: proper caching ??
      'Surrogate-Control': 'max-age: 30',
      'Surrogate-Key': surrogateKey,
      Vary: 'X-Owner,X-Repo,X-Ref,X-Repo-Root-Path',
    },
  });
}

module.exports.main = wrap(main)
  .with(status)
  .with(logger.trace)
  .with(logger);
